src/main.rs

```diff
@@ -484,34 +535,119 @@ impl BFInterp {
+                Type::IntegerType { bits: _ } => match to_type.as_ref() {
+                    Type::IntegerType { bits } => {
+                        match bits {
+                            32 => {
+                                let constant = Constant::Int {
+                                    bits: 32,
+                                    value: self.get_int_op(self.vars.get(name).unwrap()) as i32
+                                        as u64,
+                                };
+                                self.vars.insert(
+                                    dest.clone(),
+                                    ConstantOperand(ConstantRef::new(constant)),
+                                );
+                            }
+                            64 => {
+                                let constant = Constant::Int {
+                                    bits: 64,
+                                    value: self.get_int_op(self.vars.get(name).unwrap()) as i64
+                                        as u64,
+                                };
+                                self.vars.insert(
+                                    dest.clone(),
+                                    ConstantOperand(ConstantRef::new(constant)),
+                                );
+                            }
+                            _ => todo!(),
+                        };
+                    }
+                    _ => todo!(),
+                },
+                _ => todo!(),
+            },
+            ConstantOperand(_) => todo!(),
+            MetadataOperand => todo!(),
+        }
+    }
+
+    fn icmp(&mut self, pred: IntPredicate, op0: &Operand, op1: &Operand, dest: &name::Name) {
+        match op0 {
             LocalOperand { name, ty } => match ty.as_ref() {
                 Type::IntegerType { bits } => match bits {
-                    32 => match to_type.as_ref() {
-                        Type::IntegerType { bits } => {
-                            match bits {
-                                64 => {
-                                    let constant = Constant::Int {
-                                        bits: 64,
-                                        value: self.get_int_op(self.vars.get(name).unwrap()) as i32
-                                            as i64
-                                            as u64,
-                                    };
-                                    self.vars.insert(
-                                        dest.clone(),
-                                        ConstantOperand(ConstantRef::new(constant)),
-                                    );
-                                }
-                                _ => todo!(),
-                            };
-                        }
-                        _ => todo!(),
-                    },
+                    8 => todo!(),
+                    16 => todo!(),
+                    32 => {
```

Comment:
You don't need to worry about casting the values to the correct types before comparing them. Underneath they all have the same byte representation, so you can directly compare the `u64`s.

Reply:
What I meant with this is, that since comparison will work the same on the underlying `u64` value, you can get rid of this match entirely and call the same compare function for all cases, which then means you can also get rid of the generic stuff in `store_comp`.